From 2991d0670bdae496787d23160979482a292ac86a Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Sun, 8 Sep 2019 11:37:07 +0200
Subject: [PATCH 1/1] efi_loader parameter checks EFI_FILE_PROTOCOL.SetInfo()

We do not support volume label changes. No parameter checks are needed
here.

When the info for as file is changed the buffer must always contain a file
name.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 lib/efi_loader/efi_file.c | 10 ++++------
 1 file changed, 4 insertions(+), 6 deletions(-)

diff --git a/lib/efi_loader/efi_file.c b/lib/efi_loader/efi_file.c
index 71582e5149..c6190026c1 100644
--- a/lib/efi_loader/efi_file.c
+++ b/lib/efi_loader/efi_file.c
@@ -695,7 +695,9 @@ static efi_status_t EFIAPI efi_file_setinfo(struct efi_file_handle *file,
 		char *new_file_name, *pos;
 		loff_t file_size;
 
-		if (buffer_size < sizeof(struct efi_file_info)) {
+		/* The buffer will always contain a file name. */
+		if (buffer_size < sizeof(struct efi_file_info + 2) ||
+		    buffer_size < info->size) {
 			ret = EFI_BAD_BUFFER_SIZE;
 			goto out;
 		}
@@ -735,12 +737,8 @@ static efi_status_t EFIAPI efi_file_setinfo(struct efi_file_handle *file,
 		 * TODO: Support read only
 		 */
 		ret = EFI_SUCCESS;
-	} else if (!guidcmp(info_type, &efi_file_system_info_guid)) {
-		if (buffer_size < sizeof(struct efi_file_system_info)) {
-			ret = EFI_BAD_BUFFER_SIZE;
-			goto out;
-		}
 	} else {
+		/* TODO: We do not support changing the volume label */
 		ret = EFI_UNSUPPORTED;
 	}
 out:
-- 
2.20.1

